### PR TITLE
[Quant][fx][bc-breaking] Remove `remove_quant_dequant_pairs` and fix tests

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84203

Summary:
- `remove_quant_dequant_pairs` removes ops when a `quant` is followed by a `dequant`
- It looks like the quantized implementation of `layer_norm` only supports float weights, so updated the default qconfig to avoid quantizing the weight param, by attaching a placeholder parameter instead of a default_weight_parameter. This is the BC-breaking change
-  Fixes broken test, `test_norm_weight_bias`. This was the only test that broke, because the default qconfig dict we pass in quantizes the weight. I just pulled the native qconfig object and converted it to a dict.
- Adds in qconfig and backend config support for layernorm

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
```

BC Breaking Notes:

Before:
```
import torch
from torch.ao.quantization.qconfig_mapping import QConfigMapping, QConfig
from torch.ao.quantization.observer import default_weight_observer
from torch.ao.quantization.backend_config import (
    DTypeConfig,
    ObservationType,
)
from torch.ao.quantization.quantize_fx import prepare_fx, convert_fx

qconfig = QConfig(activation=qconfig.activation, weight=default_weight_observer)
qconfig_mapping = QConfigMapping().set_object_type(torch.nn.LayerNorm, q_config) \
.set_object_type(torch.nn.functional.layer_norm, q_config)

# assuming mymodel contains a LayerNorm layer or torch.nn.functional.layer_norm
m = MyModel()
example_inputs = (torch.rand(3, 3),)
m = prepare_fx(m, qconfig_mapping, example_inputs)
```

After:
```
import torch
from torch.ao.quantization.qconfig_mapping import QConfigMapping, QConfig
from torch.ao.quantization.observer import default_placeholder_observer
from torch.ao.quantization.backend_config import (
    DTypeConfig,
    ObservationType,
)
from torch.ao.quantization.quantize_fx import prepare_fx, convert_fx

qconfig = QConfig(activation=qconfig.activation, weight=default_placeholder_observer)
qconfig_mapping = QConfigMapping().set_object_type(torch.nn.LayerNorm, q_config) \
.set_object_type(torch.nn.functional.layer_norm, q_config)

# assuming mymodel contains a LayerNorm layer or torch.nn.functional.layer_norm
m = MyModel() 
example_inputs = (torch.rand(3, 3),)
m = prepare_fx(m, qconfig_mapping, example_inputs)
```

Reviewers:

Subscribers:

Tasks: Fixes https://github.com/pytorch/pytorch/issues/83110

Tags: quant, fx

Differential Revision: [D39395141](https://our.internmc.facebook.com/intern/diff/D39395141)

### CODE REVIEW

File: torch/ao/quantization/fx/convert.py [hunk 1]
```diff
@@ -795,9 +776,6 @@ def replace_observer_with_dequantize_node(node: Node, graph: Graph):
     # TODO: maybe move this to quantize_fx.py
     if not is_reference:
         model = lower_to_fbgemm(model, qconfig_map, node_name_to_scope)
-        model = remove_quant_dequant_pairs(model)
-    # TODO: this looks hacky, we want to check why we need this and see if we can
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/convert.py:799]: the todo is for next line actually :)

**Comment by jerryzh168** [comment]
there might be models being broken internally, please make sure to do ghimport and check if anything breaks

**Comment by jerryzh168** [comment]
at the high level, removing quant-dequant pair is not valid, since it changes the numerics (since quant changes numerics). the need for this pass means we did not handle some op properly in the flow so we'd need to find out which op is it

File: test/quantization/fx/test_quantize_fx.py [hunk 2]
```diff
@@ -5917,7 +5917,8 @@ def forward(self, x):
             ns.call_function(torch.quantize_per_tensor): 1,
             ns.call_function(torch.ops.quantized.linear): 1,
             ns.call_function(torch.ops.quantized.layer_norm): 1,
-            ns.call_method("dequantize"): 1,
+            ns.call_function(torch.quantize_per_channel): 1,
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:5920]: can you paste the full model before and after the change? would like to see if the final model makes sense or not

Comment (jcaip) [reply test/quantization/fx/test_quantize_fx.py:5920]: Before the change:
```python
quant type:                                                                                                                                                                                               [3/1820]
 QuantType.STATIC                                                                                                                                                                                                 
original model:                                                                                                                                                                                                   
 M(                                                                                                                                                                                                               
  (mods1): Linear()                                                                                                                                                                                               
)                                                                                                                                                                                                                 
                                                                                                                                                                                                                  
prepared model:                                                                                                                                                                                                   
 GraphModule(                                                                                                                                                                                                     
  (activation_post_process_0): HistogramObserver()                                                                                                                                                                
  (mods1): Module()                                                                                                                                                                                               
  (activation_post_process_1): PerChannelMinMaxObserver(min_val=tensor([1., 1., 1., 1., 1.]), max_val=tensor([1., 1., 1., 1., 1.]))                                                                               
  (activation_post_process_2): HistogramObserver()                                                                                                                                                                
  (activation_post_process_3): PerChannelMinMaxObserver(min_val=tensor([-0.9331, -1.3769, -1.0491, -0.2322, -1.0875]), max_val=tensor([0.3867, 1.6512, 2.4418, 0.7138, 1.4836]))                                  
  (activation_post_process_4): HistogramObserver()                                                                                                                                                                
)                                                                                                                                                                                                                 
                                                                                                                                                                                                                  
                                                                                                                                                                                                                  
                                                                                                                                                                                                                  
def forward(self, x):                                                                                                                                                                                             
    activation_post_process_0 = self.activation_post_process_0(x);  x = None                                                                                                                                      
    mods1_w = self.mods1.w                                                                                                                                                                                        
    activation_post_process_1 = self.activation_post_process_1(mods1_w);  mods1_w = None                                                                                                                          
    mods1_b = self.mods1.b                                                                                                                                                                                        
    linear = torch._C._nn.linear(activation_post_process_0, activation_post_process_1, mods1_b);  activation_post_process_0 = activation_post_process_1 = mods1_b = None                                          
    activation_post_process_2 = self.activation_post_process_2(linear);  linear = None                                                                                                                            
    scale = self.scale                                                                                                                                                                                            
    activation_post_process_3 = self.activation_post_process_3(scale);  scale = None                                                                                                                              
    bias = self.bias                                                                                                                                                                                              
    layer_norm = torch.nn.functional.layer_norm(activation_post_process_2, [5, 5], weight = activation_post_process_3, bias = bias, eps = 1e-05);  activation_post_process_2 = activation_post_process_3 = bias = 
None                                                                                                                                                                                                              
    activation_post_process_4 = self.activation_post_process_4(layer_norm);  layer_norm = None                                                                                                                           "devvm5875.lla0.facebo" 08:43 30-Aug-22
    return activation_post_process_4
     
quantized model:
 GraphModule()



def forward(self, x):
    mods1_input_scale_0 = self.mods1_input_scale_0
    mods1_input_zero_point_0 = self.mods1_input_zero_point_0
    quantize_per_tensor = torch.quantize_per_tensor(x, mods1_input_scale_0, mods1_input_zero_point_0, torch.quint8);  x = mods1_input_scale_0 = mods1_input_zero_point_0 = None
    mods1_packed_weight_0 = self.mods1_packed_weight_0
    mods1_scale_1 = self.mods1_scale_1
    mods1_zero_point_1 = self.mods1_zero_point_1
    linear = torch.ops.quantized.linear(quantize_per_tensor, mods1_packed_weight_0, mods1_scale_1, mods1_zero_point_1);  quantize_per_tensor = mods1_packed_weight_0 = mods1_scale_1 = mods1_zero_point_1 = None
    scale = self.scale
    bias = self.bias
    _scale_1 = self._scale_1
    _zero_point_1 = self._zero_point_1
    layer_norm_1 = torch.ops.quantized.layer_norm(linear, [5, 5], weight = scale, bias = bias, eps = 1e-05, output_scale = _scale_1, output_zero_point = _zero_point_1);  linear = scale = bias = _scale_1 = _zero
_point_1 = None
    dequantize_4 = layer_norm_1.dequantize();  layer_norm_1 = None
    return dequantize_4
 ```

Comment (jcaip) [reply test/quantization/fx/test_quantize_fx.py:5920]: After the change:
```python
quant type:                                                                                                                                                                                                                                                          
 QuantType.STATIC                                                                                                                                                                                                                                                    
original model:                                                                                                                                                                                                                                                      
 M(                                                                                                                                                                                                                                                                  
  (mods1): Linear()                                                                                                                                                                                                                                                  
)                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                     
prepared model:                                                                                                                                                                                                                                                      
 GraphModule(                                                                                                                                                                                                                                                        
  (activation_post_process_0): HistogramObserver()                                                                                                                                                                                                                   
  (mods1): Module()                                                                                                                                                                                                                                                  
  (activation_post_process_1): PerChannelMinMaxObserver(min_val=tensor([1., 1., 1., 1., 1.]), max_val=tensor([1., 1., 1., 1., 1.]))                                                                                                                                  
  (activation_post_process_2): HistogramObserver()                                                                                                                                                                                                                   
  (activation_post_process_3): PerChannelMinMaxObserver(min_val=tensor([-0.9331, -1.3769, -1.0491, -0.2322, -1.0875]), max_val=tensor([0.3867, 1.6512, 2.4418, 0.7138, 1.4836]))                                                                                     
  (activation_post_process_4): HistogramObserver()                                                                                                                                                                                                                   
)                                                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                     
def forward(self, x):                                                                                                                                                                                                                                                
    activation_post_process_0 = self.activation_post_process_0(x);  x = None                                                                                                                                                                                         
    mods1_w = self.mods1.w                                                                                                                                                                                                                                           
    activation_post_process_1 = self.activation_post_process_1(mods1_w);  mods1_w = None                                                                                                                                                                             
    mods1_b = self.mods1.b                                                                                                                                                                                                                                           
    linear = torch._C._nn.linear(activation_post_process_0, activation_post_process_1, mods1_b);  activation_post_process_0 = activation_post_process_1 = mods1_b = None                                                                                             
    activation_post_process_2 = self.activation_post_process_2(linear);  linear = None                                                                                                                                                                               
    scale = self.scale                                                                                                                                                                                                                                               
    activation_post_process_3 = self.activation_post_process_3(scale);  scale = None                                                                                                                                                                                 
    bias = self.bias                                                                                                                                                                                                                                                 
    layer_norm = torch.nn.functional.layer_norm(activation_post_process_2, [5, 5], weight = activation_post_process_3, bias = bias, eps = 1e-05);  activation_post_process_2 = activation_post_process_3 = bias = None                                               
    activation_post_process_4 = self.activation_post_process_4(layer_norm);  layer_norm = None                                                                                                                                                                       
    return activation_post_process_4                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                     
quantized model:                                                                                                                                                                                                                                                     
 GraphModule()                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                     
def forward(self, x):                                                                                                                                                                                                                                                
    mods1_input_scale_0 = self.mods1_input_scale_0                                                                                                                                                                                                                   
    mods1_input_zero_point_0 = self.mods1_input_zero_point_0                                                                                                                                                                                                         
    quantize_per_tensor = torch.quantize_per_tensor(x, mods1_input_scale_0, mods1_input_zero_point_0, torch.quint8);  x = mods1_input_scale_0 = mods1_input_zero_point_0 = None                                                                                      
    mods1_packed_weight_0 = self.mods1_packed_weight_0                                                                                                                                                                                                               
    mods1_scale_1 = self.mods1_scale_1                                                                                                                                                                                                                               
    mods1_zero_point_1 = self.mods1_zero_point_1                                                                                                                                                                                                                     
    linear = torch.ops.quantized.linear(quantize_per_tensor, mods1_packed_weight_0, mods1_scale_1, mods1_zero_point_1);  quantize_per_tensor = mods1_packed_weight_0 = mods1_scale_1 = mods1_zero_point_1 = None                                                     
    scale = self.scale                                                                                                                                                                                                                                               
    _scale_0 = self._scale_0                                                                                                                                                                                                                                         
    _zero_point_0 = self._zero_point_0                                                                                                                                                                                                                               
    quantize_per_channel_1 = torch.quantize_per_channel(scale, _scale_0, _zero_point_0, 0, torch.qint8);  scale = _scale_0 = _zero_point_0 = None                                                                                                                    
    dequantize_3 = quantize_per_channel_1.dequantize();  quantize_per_channel_1 = None                                                                                                                                                                               
    bias = self.bias                                                                                                                                                                                                                                                 
    _scale_1 = self._scale_1                                                                                                                                                                                                                                         
    _zero_point_1 = self._zero_point_1                                                                                                                                                                                                                               
    layer_norm_1 = torch.ops.quantized.layer_norm(linear, [5, 5], weight = dequantize_3, bias = bias, eps = 1e-05, output_scale = _scale_1, output_zero_point = _zero_point_1);  linear = dequantize_3 = bias = _scale_1 = _zero_point_1 = None                      
    dequantize_4 = layer_norm_1.dequantize();  layer_norm_1 = None                                                                                                                                                                                                   
    return dequantize_4   
```

**Comment by jcaip** [comment]
@jcaip has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39144610).

File: test/quantization/fx/test_quantize_fx.py (hunk 2 above)
Comment (jerryzh168) [reply test/quantization/fx/test_quantize_fx.py:5920]: I think there are some problems with the lowering code for quantized layer_norm, we need to fix that, basically the lowering does did not fuse the dequantize into the quantized layer_norm op (only activation was fused), the fusion pattern should be:
```
weight - quant - dequant \
input - quant - dequant - layer_norm - quant - dequant
```

Comment (jerryzh168) [reply test/quantization/fx/test_quantize_fx.py:5920]: > Gotcha, are you talking about dequantize_3?

Yes.

> Is the expected behavior to be able to pass in the quantized scale directly to quantized.layer_norm?

I guess the weight of layernorm is not quantized: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp#L2672

that means we need to skip adding q/dq for weight, this requires changes to both the qconfig for layer_norm and the BackendConfig entry for layernorm

Comment (jcaip) [reply test/quantization/fx/test_quantize_fx.py:5920]: @jerryzh168 Does this look right?

```python

quant type:
 QuantType.STATIC
original model:
 M(
  (mods1): Linear()
)

prepared model:
 GraphModule(
  (activation_post_process_0): HistogramObserver()
  (mods1): Module()
  (activation_post_process_1): PerChannelMinMaxObserver(min_val=tensor([1., 1., 1., 1., 1.]), max_val=tensor([1., 1., 1., 1., 1.]))
  (activation_post_process_2): HistogramObserver()
  (activation_post_process_3): PlaceholderObserver()
  (activation_post_process_4): HistogramObserver()
)



def forward(self, x):
    activation_post_process_0 = self.activation_post_process_0(x);  x = None
    mods1_w = self.mods1.w
    activation_post_process_1 = self.activation_post_process_1(mods1_w);  mods1_w = None
    mods1_b = self.mods1.b
    linear = torch._C._nn.linear(activation_post_process_0, activation_post_process_1, mods1_b);  activation_post_process_0 = activation_post_process_1 = mods1_b = None
    activation_post_process_2 = self.activation_post_process_2(linear);  linear = None
    scale = self.scale
    activation_post_process_3 = self.activation_post_process_3(scale);  scale = None
    bias = self.bias
    layer_norm = torch.nn.functional.layer_norm(activation_post_process_2, [5, 5], weight = activation_post_process_3, bias = bias, eps = 1e-05);  activation_post_process_2 = activation_post_process_3 = bias = None
    activation_post_process_4 = self.activation_post_process_4(layer_norm);  layer_norm = None
    return activation_post_process_4
    
# To see more debug info, please use `graph_module.print_readable()`

quantized model:
 GraphModule()



def forward(self, x):
    mods1_input_scale_0 = self.mods1_input_scale_0
    mods1_input_zero_point_0 = self.mods1_input_zero_point_0
    quantize_per_tensor = torch.quantize_per_tensor(x, mods1_input_scale_0, mods1_input_zero_point_0, torch.quint8);  x = mods1_input_scale_0 = mods1_input_zero_point_0 = None
    mods1_packed_weight_0 = self.mods1_packed_weight_0
    mods1_scale_1 = self.mods1_scale_1
    mods1_zero_point_1 = self.mods1_zero_point_1
    linear = torch.ops.quantized.linear(quantize_per_tensor, mods1_packed_weight_0, mods1_scale_1, mods1_zero_point_1);  quantize_per_tensor = mods1_packed_weight_0 = mods1_scale_1 = mods1_zero_point_1 = None
    scale = self.scale
    bias = self.bias
    _scale_0 = self._scale_0
    _zero_point_0 = self._zero_point_0
    layer_norm_1 = torch.ops.quantized.layer_norm(linear, [5, 5], weight = scale, bias = bias, eps = 1e-05, output_scale = _scale_0, output_zero_point = _zero_point_0);  linear = scale = bias = _scale_0 = _zero_point_0 = None
    dequantize_3 = layer_norm_1.dequantize();  layer_norm_1 = None
    return dequantize_3
    
# To see more debug info, please use `graph_module.print_readable()`
'placeholder' 'x' 'x' () {}
'get_attr' 'mods1_input_scale_0' 'mods1_input_scale_0' () {}
'get_attr' 'mods1_input_zero_point_0' 'mods1_input_zero_point_0' () {}
'call_function' 'quantize_per_tensor' <built-in method quantize_per_tensor of type object at 0x7f9ea0a13180> (x, mods1_input_scale_0, mods1_input_zero_point_0, torch.quint8) {}
'get_attr' 'mods1_packed_weight_0' 'mods1_packed_weight_0' () {}
'get_attr' 'mods1_scale_1' 'mods1_scale_1' () {}
'get_attr' 'mods1_zero_point_1' 'mods1_zero_point_1' () {}
'call_function' 'linear' <OpOverloadPacket(op='quantized.linear')> (quantize_per_tensor, mods1_packed_weight_0, mods1_scale_1, mods1_zero_point_1) {}
'get_attr' 'scale' 'scale' () {}
'get_attr' 'bias' 'bias' () {}
'get_attr' '_scale_0' '_scale_0' () {}
'get_attr' '_zero_point_0' '_zero_point_0' () {}
'call_function' 'layer_norm_1' <OpOverloadPacket(op='quantized.layer_norm')> (linear, [5, 5]) {'weight': scale, 'bias': bias, 'eps': 1e-05, 'output_scale': _scale_0, 'output_zero_point': _zero_point_0}
'call_method' 'dequantize_3' 'dequantize' (layer_norm_1,) {}
'output' 'output' 'output' (dequantize_3,) {}
```

Comment (jerryzh168) [reply test/quantization/fx/test_quantize_fx.py:5920]: yeah this looks good

**Comment by jerryzh168** [comment]
don't you need to change the dtype config for layer_norm? https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/backend_config/_common_operator_config_utils.py#L370 maybe there is some problem with the prepare code

**Comment by jerryzh168** [comment]
currently it is: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/backend_config/native.py#L96, and this is saying the weight should have dtype qint8 I think

**Comment by jcaip** [comment]
@jerryzh168 that dtype only specifies input/output as qint8 so I think that's why it works. 
```
default_op_quint8_dtype_config = DTypeConfig(
    input_dtype=torch.quint8,
    output_dtype=torch.quint8,
)
```
I can make a custom one and make it more explicit if that is better.

**Comment by jerryzh168** [comment]
> @jerryzh168 that dtype only specifies input/output as qint8 so I think that's why it works.
> 
> ```
> default_op_quint8_dtype_config = DTypeConfig(
>     input_dtype=torch.quint8,
>     output_dtype=torch.quint8,
> )
> ```
> 
> I can make a custom one and make it more explicit if that is better.

I see, I feel we should have another dtypeconfig for these ops and specify the dtype for weight as well

File: torch/ao/quantization/backend_config/native.py [hunk 3]
```diff
@@ -61,6 +62,14 @@
     is_dynamic=True,
 )
 
+# Needed for LayerNorm and f.layer_norm, since currently the kernel only supports float weights
+input_output_only_quint8_dtype_config = DTypeConfig(
+    input_dtype=torch.quint8,
+    output_dtype=torch.quint8,
+    weight_dtype=torch.float,
+    bias_dtype=torch.float,
```
Comment (jerryzh168) [inline torch/ao/quantization/backend_config/native.py:70]: just thought about this again, looks like by default we assume the weight dtype is supported: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py#L190 maybe the default should be float/not quantized?

Comment (jerryzh168) [reply torch/ao/quantization/backend_config/native.py:70]: basically when weight_dtype is not specified in BackendConfig it is None, so the check for weight_dtype specified by QConfig will always pass right now, this doesn't sound correct I feel

**Review by jerryzh168 — APPROVED** [review]
looks good, thanks for the fixes. Maybe we can follow up with a PR that fix the default behavior of weight_dtype and bias_dtype

**Comment by jerryzh168** [comment]
btw test plan can be more specific, like:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
python test/test_quantization.py TestQuantizeFxModels
```

**Comment by jerryzh168** [comment]
also please make sure internal tests are all passing, I suspect there might be tests failing internally

**Comment by jcaip** [comment]
@jcaip has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39395141).

File: torch/ao/quantization/qconfig_mapping.py [hunk 4]
```diff
@@ -73,6 +74,10 @@ def _get_default_qconfig_mapping(is_qat: bool, backend: str, version: int) -> QC
     else:
         qconfig_transpose = qconfig
 
+    # currently layernorm only supports float weights
+    # we have to add this because otherwise there will be a extra quantize-dequantize pair
+    qconfig_layernorm = QConfig(activation=qconfig.activation, weight=default_placeholder_observer)
```
Comment (jerryzh168) [inline torch/ao/quantization/qconfig_mapping.py:79]: @jcaip for the BC breaking notes, we just need to mention the qconfig for layer_norm, basically we can say something like:

Before:
qconfig_layer_norm = get_default_qconfig("fbgemm")

After:
qconfig_layer_norm = QConfig(activation=default_observer, weight=default_placeholder_observer)

**Pull request closed without merging**